### ISSUE TITLE
Insert NULL pgtype.JSONB: cannot encode status undefined

### ISSUE THREAD

**Issue by arigon** [jackc/pgx#952]
I am trying to insert an undefined JSONB field but it fails with _cannot encode status undefined_.

If I change the field to *pgtype.JSONB the error changes to: _value method github.com/jackc/pgtype.JSONB.PreferredParamFormat called using nil *JSONB pointer_
I thought it would be dereferenced automatically?

Shouldn't undefined fields always be handled as NULL?

**Comment by jackc** [comment]
The pgtype library as a whole distinguishes between undefined and null. This was so ORMs could insert a structure and easily distinguish between "set column to null" and "do not change column". In hindsight that may not have been the best idea but it is far to late to change in a non-major release.

**Comment by arigon** [comment]
@jackc is there an alternative inside pgx? As a workaround I implemented my own dereferencing and sanitizing functions for my microservices.

**Comment by jackc** [comment]
Well, the pgx / pgtype is pluggable so you can replace the default handlers for any type if you really want to. But my guess is that you don't need to use *pgtype.JSONB directly at all. You can use types like `string`, `[]byte`, and `map[string]interface{}` directly with jsonb types.

**Comment by califlower** [comment]
I was actually about to open this issue myself and was pretty surprised to see it open. 

This change is pretty disappointing. I used to use JSONB fine, but now with this change it's requiring a lot of extra work to change stuff.

I never want to be the one to criticize an open source project too much, as I know it's volunteer work and this library is overall incredible, but I think the current behavior of failing on nil fields is pretty footgunny

**Comment by jackc** [comment]
> This change is pretty disappointing.

Maybe I don't understand the original issue then. To my knowledge the current behavior has been the same since json/jsonb support was introduced.

Can one of you provide a code example?

**Comment by califlower** [comment]
Sorry for the late response! @jackc 

So essentially the issue is, in previous versions using the JsonB returned by a database struct and then converting that jsonB into a typed Golang struct would work properly for cases where the underlying response from the DB was null for the jsonb column.

Now it will throw a method github.com/jackc/pgtype.JSONB.PreferredParamFormat called using nil *JSONB pointer because all calls to (Unmarshal? I think) call JSONB.PreferredParamFormat on the null response and then it fails.

On the other hand, I took your advice to heart and converted the json struct fields to a string type which has worked wonderfully for my application. 

I didn't realize JSONB is more of an internal library type

**Comment by jackc** [comment]
> On the other hand, I took your advice to heart and converted the json struct fields to a string type which has worked wonderfully for my application.

> I didn't realize JSONB is more of an internal library type

Right, it is rarely needed to use the `pgtype.JSONB` type directly. And also, beyond just scanning json values into a string you can directly scan them into the struct you will unmarshal the json into.

**Comment by ebo-bpi** [comment]
@jackc I'm using pgx in combination with sqlc and this is working wonderfull! all types are created base on the schema structure. also the pgtype.JSON is working fine. only problem I have is with nil/null values. mostly when inserting to DB I'm getting this error. an empty "{}"  object solves this issue, but I find this a workaround and I need to thing of this in every service / function. It seems only with JSON types a null insert is failing. What do you think? How can we tackle this?

**Comment by jackc** [comment]
@ebo-bpi Maybe I still don't really understand the issue. `pgtype.Undefined` is the same in all types, not just `pgtype.JSON[B]`.

FWIW, in the future this shouldn't be an issue. `v5` is removing the `Status` field altogether. Types will instead have a `Valid` boolean like `database/sql` types.

**Comment by ebo-bpi** [comment]
Let me explain my situation, maybe it becomes more clear then.
I'm using SQLC to generate the sql code needed. sqlc is using pg types and for almost everything it is working fine. 
now that is happening with some types allowing null in the db. varchar for example becomes sql.nullstring. same for bigint and time. for JSON there is no nullJSON type, and I think that is where the issue comes. when a JSON field in not specified (so an undefined/ nil pointer) and when you try to insert a row to the DB it will cause the error. 
This is not happening with the other types. 
hope this explains it a bit.

**Comment by thanishsid** [comment]
> Let me explain my situation, maybe it becomes more clear then.
> I'm using SQLC to generate the sql code needed. sqlc is using pg types and for almost everything it is working fine. 
> now that is happening with some types allowing null in the db. varchar for example becomes sql.nullstring. same for bigint and time. for JSON there is no nullJSON type, and I think that is where the issue comes. when a JSON field in not specified (so an undefined/ nil pointer) and when you try to insert a row to the DB it will cause the error. 
> This is not happening with the other types. 
> hope this explains it a bit.

Could you try constructing your sqlc query with a nullif ex: INSERT INTO your_table VALUES (
NULLIF(@your_json_param::JSON, '{}')
...other columns inserted
);

**Comment by ebo-bpi** [comment]
That would make no difference as the type is still pgtype.JSON and  issue is not at the db, but before that. 
Actually as a sort of workaround I indeed send an empty json object by default. this is a bit ugly, but working for now.

**Comment by thanishsid** [comment]
> That would make no difference as the type is still pgtype.JSON and  issue is not at the db, but before that. 
> Actually as a sort of workaround I indeed send an empty json object by default. this is a bit ugly, but working for now. 
> 

But if you're sending an empty json object the value wouldn't be null in the db it will be an empty json right ?

**Comment by arigon** [comment]
This is my workaround which works flawless for over a year now. This function is called before every insert and update.

```
func dereferencePgtypes(args ...interface{}) []interface{} {
	var newArgs []interface{}
	for _, arg := range args {
		// JSONB pointer cannot be inserted by the DB driver => dereference
		if v, ok := arg.(*pgtype.JSONB); ok {
			refVal := reflect.ValueOf(v)
			if refVal.Kind() == reflect.Ptr {
				if refVal.IsZero() {
					arg = pgtype.JSONB{}
				} else {
					arg = refVal.Elem().Interface()
				}
			}
		}
		if v, ok := arg.(pgtype.JSONB); ok {
			if v.Status == pgtype.Undefined {
				v.Status = pgtype.Null
			}
			newArgs = append(newArgs, v)
		} else if v, ok := arg.(pgtype.Timestamptz); ok {
			if v.Status == pgtype.Undefined {
				v.Status = pgtype.Null
			}
			newArgs = append(newArgs, v)
		} else {
			newArgs = append(newArgs, arg)
		}
	}
	return newArgs
}
```

**Comment by D1kz** [comment]
Hello, folks!
Facing the same issue here. Thank you for this thread and I completely understand that `pgtype.JSONB` is for internal use and I need to refactor our code or use|find|implement a workaround on that.

Sorry for bothering you, @jackc. But could you tell when `v5` is coming? (i saw the milestone here, but there is no due date in there :))

**Comment by jackc** [comment]
@D1kz pgx v5 should be released this month (Sept. 2022).

**Comment by kirk-anchor** [comment]
When using the pgx native interface on pgx-v4.18.1 and pgtype-v1.14.0, using an empty `pgtype.JSONB{}` in a composite type results in the column having a value of `'0'::jsonb` on Postgres. I'm not sure if it's because of the `Get()` method returning the status field instead of nil but this seems like a bug. I think it should return an error instead of using `'0'::jsonb` https://github.com/jackc/pgtype/blob/cb24fcc9879f1b0e41060f790146302a9af756e1/json.go#L76

**Comment by rubiagatra** [comment]
Do you know if this issue is still relevant in V5? @kirk-anchor @arigon ?
